advanced_alchemy/types/file_object.py

```diff
@@ -0,0 +1,313 @@
+    etag: str | None | EmptyType = Empty
+    version_id: str | None | EmptyType = Empty
+
+    def to_dict(self) -> dict[str, Any]:
+        """Convert metadata to dictionary."""
+        return simple_asdict(self, exclude_empty=True)
+
+    @classmethod
+    def from_dict(cls, data: dict[str, Any]) -> FileMetadata:
+        """Create metadata from dictionary."""
+        return cls(**data)
+
+
+@runtime_checkable
+class StorageBackend(Protocol):
+    """Protocol for storage backend implementations."""
+
+    async def save_file(self, path: str, data: bytes, content_type: str | None = None) -> None:
+        """Save file data to storage.
+
+        Args:
+            path: Path where to store the file
+            data: File data to store
+            content_type: Optional MIME type of the file
+        """
+        ...
+
+    async def get_url(self, path: str, expires_in: int) -> str:
+        """Get access URL for file.
+
+        Args:
+            path: Path to the file
+            expires_in: Expiration time in seconds
+
+        Returns:
+            Presigned URL for file access
+        """
+        ...
+
+    async def get_upload_url(
+        self,
+        path: str,
+        expires_in: int,
+        content_type: str,
+        max_size: int | None = None,
+    ) -> str:
+        """Get upload URL for file.
+
+        Args:
+            path: Path where the file will be stored
+            expires_in: Expiration time in seconds
+            content_type: MIME type of the file
+            max_size: Optional maximum file size in bytes
+
+        Returns:
+            Presigned URL for file upload
+        """
+        ...
+
+    async def delete_file(self, path: str) -> None:
+        """Delete a file from storage.
+
+        Args:
+            path: Path to the file to delete
+        """
+        ...
+
+
+class FSSpecBackend(StorageBackend):
+    """FSSpec implementation of storage backend."""
+
+    def __init__(self, backend: str, base_path: str = "", **options: Any) -> None:
+        """Initialize FSSpec backend.
+
+        Args:
+            backend: Storage backend to use (e.g., 'file', 'gcs', 's3')
+            base_path: Base path/prefix for stored files
+            **options: Additional options for fsspec
+
+        Raises:
+            MissingDependencyError: If fsspec is not installed
+        """
+        if fsspec is None:
+            raise MissingDependencyError(package="fsspec", install_package="fsspec")
+
+        self.backend = backend
+        self.base_path = base_path.rstrip("/")
+        self._fs = fsspec.filesystem(backend, **options)
+
+    async def save_file(self, path: str, data: bytes, content_type: str | None = None) -> None:
+        """Save file data using FSSpec."""
+        full_path = f"{self.base_path}/{path}".lstrip("/")
+
+        if hasattr(self._fs, "async_put"):
+            await self._fs.async_put(full_path, data)
+        else:
+            self._fs.put(full_path, data)
```

Comment:
This should run in a threadpool so it doesn't block

Reply:
Cool. Let me know what else you see.  This is still very much a draft.  

I think I need to refactor the API a bit before it's ready for actual testing.